.changeset/tender-pens-sell.md

```diff
@@ -0,0 +1,46 @@
+---
+"astro": minor
+---
+
+Adds a 2 news option to the Integrations API `astro:config:setup` hook: `codegenDir` and `injectDts`. 
+
+Integrations authors can now interact with the `.astro` directory more easily! Its path is now available as `codegenDir` (of type `URL`) inside of `astro:config:setup`:
+
+```ts
+import type { AstroIntegration } from "astro"
+
+export function integration(): AstroIntegration {
+    return {
+        name: "my-integration",
+        hooks: {
+            "astro:config:setup": ({ codegenDir }) => {
+                const filePath = new URL("./data.json", codegenDir)
+            }
+        }
+    }
+}
+```
+
+The `.astro` directory is now created before any integration runs so you don't have to anymore!
+
+To make things even easier, a new utility is now available in `astro:config:setup` as well: `injectDts`! It allows you to add `.d.ts` files inside your users project and it will be automatically included.
+
+```ts
+import type { AstroIntegration } from "astro"
+
+export function integration({ param }: { param: string }): AstroIntegration {
+    return {
+        name: "my-integration",
+        hooks: {
+            "astro:config:setup": ({ injectDts }) => {
+                injectDts({
+                    name: "my-integration.d.ts",
+                    content: `declare module "virtual:my-integration/config" {
+                        export const param: string;
+                    }`
+                })
```

Comment:
I have some reservations about this API. I don't think integrations should be allowed to specify a file name because there isn't a need:
- they might end up creating folders e.g. `folder/folder/folder/file.d.ts`. We don't want that.
- Astro should own file system operations, even file names
- with this approach, integrations can decide to change file names from one version to another, and the old files might end up there forever, unused

Reply:
If we go with my suggestion on Blu's comment, that would solve this issue. And to also address Fry naming concerns, it could look like this for integrations:
```ts
injectTypes(`declare module ....`)
```
That would always create a file at `${root}/.astro/integrations/${integrationName}/types.d.ts`

Reply:
Actually, yesterday i was working on an integration that required several dts files, to avoid conflicts. So I wonder how that would go with such API 🤔

Reply:
Ideally, the API will save the string inside an array. At the end everything will go inside the same file

Reply:
I've made more testing and there are cases where several dts are required because stuff would conflict between them. For example:
```ts
declare module "virtual:foo" {
	export const bar: string;
}

import type _i18next0 from "../src/locales/en/test.json";
declare module "i18next" {
  interface CustomTypeOptions {
    defaultNS: "test";
    resources: {
      test: typeof _i18next0;
    };
  }
}
```
here because there's an import, the first module declaration is broken. I've tried many things but it can't work without being 2 files. So I think it could be useful to be able to create several dts per integration

Reply:
That's... unsettling. But as long as we control the name of the files, I believe it's fine and we can work with this requirement

Reply:
You need to support multiple declarations files anyway because you need to be able to output both ambient and non-ambient types.